swan-lake/learn/writing-tests.md

```diff
@@ -0,0 +1,345 @@
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertIntEquals() {
+
+    int answer = 0;
+    int a = 5;
+    int b = 3;
+    answer = intAdd(a, b);
+    test:assertNotEquals(answer, 8, msg = "Matches");
+}
+
+function intAdd(int a, int b) returns (int) {
+    return (a + b);
+}
+```
+
+### **assertFail(string message)**
+
+Fails the test. Useful when we want to fail a test while in execution based on a check for a condition.
+
+***Example:***
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function foo() {
+    error? e = trap bar(); // Expecting `bar()` to panic
+    if (e is error) {
+        test:assertEquals(e.reason(), "Invalid Operation", msg = "Invalid error reason"); // Some other assertions
+    } else {
+        test:assertFail(msg = "Expected an error");
+    }
+}
+```
+
+## Setup and Teardown
+
+Following test annotations can be used for setup and teardown instructions. These annotations enable executing instructions in different levels.
+
+### @test:BeforeSuite {}
+
+The function specified following the annotation will be run once before any of the tests in the test suite is run. This can be used for initializing test suite level aspects.
+
+***Example:***
+
+```ballerina
+import ballerina/io;
+import ballerina/test;
+
+// The `BeforeSuite` function is executed before all test functions in this module. 
+@test:BeforeSuite
+function beforeFunc() {
+    io:println("I'm the before suite function!");
+}
+
+// Test function.
+@test:Config {}
+function testFunction1() {
+    io:println("I'm in test function 1!");
+    test:assertTrue(true, msg = "Failed");
+}
+
+// Test function.
+@test:Config {}
+function testFunction2() {
+    io:println("I'm in test function 2!");
+    test:assertTrue(true, msg = "Failed");
+}
+```
+
+### @test:BeforeEach {}
+
+The function specified following the annotation will be run before every test within the test suite is run. This can be used for repeatedly initializing test level aspects before every test function.
+
+***Example:***
+
+```ballerina
+import ballerina/io;
+import ballerina/test;
+
+// Before each function, which is executed before each test function
+@test:BeforeEach
+function beforeFunc() {
+    io:println("I'm the before function!");
+}
+
+// Test function
+@test:Config {}
+function testFunction1() {
+    io:println("I'm in test function 1!");
+    test:assertTrue(true, msg = "Failed!");
+}
+
+// Test function
+@test:Config {}
+function testFunction2() {
+    io:println("I'm in test function 2!");
+    test:assertTrue(true, msg = "Failed!");
+}
+
+// Test function
+@test:Config {}
+function testFunction3() {
+    io:println("I'm in test function 3!");
+    test:assertTrue(true, msg = "Failed!");
+}
+```
+
+### @test:AfterSuite {}
+
+The function specified following the annotation will be run once after all of the tests in the test suite is run. This can be used for cleaning up test suite level aspects. The test suite covers tests related to a module.
```

Comment:
```suggestion
The function specified following the annotation will be run once after all of the tests in the test suite is run. This can be used for cleaning up the test suite level aspects. The test suite covers tests related to a module.
```